doc/man/katedec.1.in
.\" Copyright (C) 2008 Vincent Penquerc'h.
.\" This file is part of the Kate codec library.
.\" Written by Vincent Penquerc'h.
.\"
.\" Use, distribution and reproduction of this library is governed
.\" by a BSD style source license included with this source in the
.\" file 'COPYING'. Please read these terms before distributing. */


.TH "KATEDEC" 1 "01 september 2008" "libkate @VERSION@ "libkate documentation"

.SH NAME
katedec \- decodes Kate streams to a textual description

.SH SYNOPSIS
.B katedec
.B "[\-hVv]"
.BI "[\-o " outfile ]
.BI "[\-t " type ]
[ infile ]

.SH DESCRIPTION
.B katedec
decodes Kate streams to libkate's custom description language.

.SH OPTIONS
.TP
\fB\-h\fR
Show command line help.
.TP
\fB\-V\fR
Show version information.
.TP
\fB\-v\fR
Increase verbosity.
.TP
\fB\-o\fR outfile
Write the output description file to the given file name (writes to stdout if not specified).
It is possible to write to multiple files (eg, if the input contains several Kate streams)
by passing printf style formatting tags in the filename. Formatting tags are strings starting
with a '%' character, followed by the tag itself. Known tags are:
  %    replaced by a verbatim % character
  l    replaced by the stream language
  c    replaced by the stream category
  s    replaced by the stream serial number, in hexadecimal format
  i    replaced by the index of the stream in the input file, starting from 0

.TP
\fB\-t\fR type
Set the output format type:
    kate: Kate description format
    srt: SubRip format (non text information will be lost)

.SH EXAMPLES

Decodes a Kate stream to a textual description:

    katedec \-o output.kate input.ogg

Decodes several multiplexed Kate streams to SubRip files:

    katedec \-t srt \-o output.%l.kate input.ogg

.SH SEE ALSO
.PP
\fBkateenc\fR(1), \fBkatalyzer\fR(1)



ChangeLog
0.3.0
  support for repeats, to help seeking
  kate_encode_set_id is now obsolete, an event's id is now autogenerated
  fix kate_granule_time ignoring the time offset
  expose kate_granule_split_time in the API

0.2.8 - 14 december 2008
  kateenc can now read an event's text from an external file
  new code in the UTF-8 decoder to protect against overlong sequences
  add more tests to make check

0.2.7 - 24 november 2008
  fix wildly embarassing bug - forgot to copy data in kate_packet_init
  fix size mixup when removing markup at decode time
  add a couple tests for make check
  add extra validation checks to the decoder

0.2.6 - 14 november 2008
  fix a katedec crash if decoding more than one Kate stream to stdout
  various minor cleanups

0.2.5 - 16 october 2008
  fix VPATH documentation build (yselkowitz/Cygwin Ports project)
  uninstalled pc files now link against static libs rather than libtool ones

0.2.4 - 13 october 2008
  support for both hexadecimal and decimal numeric entities
  katedec can now output LRC files
  kateenc now supports variable timestamping precision in LRC files
  more tweaks to try to find a suitable type for kate_int64_t

0.2.3 - 06 october 2008
  numeric entities are now interpreted as decimal, rather than hexadecimal
  kate_info_matches_language now does better exact/partial matching
  try harder to define kate_int64_t if included by C89 code not using autoconf
  use kate_uintptr_t in kate/kate.h, fixes build when uintptr_t is not found
  kate_info_set_language now truncates according to RFC 4646
  kate_info_set_language does some basic tag validation
  more error checking in the tools

0.2.2 - 02 october 2008
  fix several memory leaks on error code paths
  fix build when doxygen isn't found (Alexis Ballier)

0.2.1 - 29 september 2008
  generate correct filenames for absolute paths
  build fix for Cygwin/Mingw32 (yselkowitz/Cygwin Ports project)
  minor code simplification in the tracker

0.2.0 - 22 september 2008
  bitstream 0.4
  glyph pointers and markers can now have a custom animated image
  bitmaps now have a logical position (x/y offset)
  better height control for karaoke use of the glyph pointer
  karaoke timing info is now extracted from enhanced lrc files
  line width can now be controlled (via a motion)
  add text wrapping property to style
  comments are now validated against Vorbis comment rules
  more validity checks in the encoder and API

0.1.12 - 09 september 2008
  granule rate numerator and denominator are now unsigned
  kateenc can be given comments on the command line to add to the output stream
  some improvements in srt parsing
  katedec now writes srt times using the more widespread ',' instead of '.'
  build system improvements

0.1.11 - 04 september 2008
  fix version number being truncated

0.1.10 - 03 september 2008
  katedec does not unlink output files for no good reason anymore :)

0.1.9 - 02 september 2008
  defend against corrupt streams and pathological input
  fix some encoding properties not being reset for each event
  katedec can now decode multiple Kate streams in an Ogg stream
  katedec can now output srt format files
  fix make check on PPC archs

0.1.8 - 25 august 2008
  various build fixes for particular system configurations
  new -c (category) option to kateenc
  new man pages for kateenc and katedec

0.1.7 - 11 august 2008
  autotools support
  tools renamed to kateenc and katedec, and now installed
  win32 fixes (Roman Vorobets)
  win32 build files (Roman Vorobets)
  fix off-by-ten tracker scaling bug
  the encoder now reads files from the input file directory, if any

0.1.6 - 21 july 2008
  new kate_high_decode_get_comments api function
  libraries can now build in C89 mode
  allow building without libogg (liboggkate and tools will not be built)
  the bitstream format is now fully documented
  use new names for oggz tools
  the encoder can now read .lrc (lyrics) files

0.1.5 - 08 july 2008
  bitstream 0.3
  new kate_encode_get_granule api function
  space metrics can now be expressed in millionths
  the info header can now hold the original canvas size
  fix misguided error throwing if a reserved area is not 0

0.1.4 - 23 june 2008
  Kate magic is now 8 bytes (though the bitstream is unchanged)
  fix an EOS packet encoding bug when using keepalive packets

0.1.3 - 02 june 2008
  remove last dependency on libogg from libkate (the bitpacker)
  new kate_region_init, kate_style_init, etc, functions
  RLE and delta compression for paletted images
  fix 64 bit warnings

0.1.2 - 15 may 2008
  bitstream 0.2
  background image
  text can now be escaped for markup if needed
  all PNG images can now be read as raw images
  new style attributes: font name, justify
  new region attributes: clip
  new mapping: bitmap size
  pkg-config files moved to misc/pkgconfig
  dodgy workarounds for building with enforcing SELinux

0.1.1 - 31 march 2008
  Examples now build with the new API
  Several fixes for bugs found with valgrind

0.1.0 - 18 march 2008
  bitstream 0.1
  First released version


tools/katedec.c
/* Copyright (C) 2008 Vincent Penquerc'h.
   This file is part of the Kate codec library.
   Written by Vincent Penquerc'h.

   Use, distribution and reproduction of this library is governed
   by a BSD style source license included with this source in the
   file 'COPYING'. Please read these terms before distributing. */


#ifdef HAVE_CONFIG_H
#include "config.h"
#endif
#ifdef HAVE_UNISTD_H
#include <unistd.h>
#endif
#include <stdarg.h>
#include <stdio.h>
#ifdef HAVE_STDLIB_H
#include <stdlib.h>
#endif
#include <errno.h>
#ifdef HAVE_STRING_H
#include <string.h>
#endif
#include <ctype.h>
#include <ogg/ogg.h>
#include "kate/oggkate.h"
#include "kate_internal.h"
#include "ksrt.h"
#include "klrc.h"
#include "kkate.h"
#include "kfuzz.h"
#include "kstream.h"
#include "kread.h"
#include "kutil.h"

static void print_version(void)
{
  printf("Kate reference decoder - %s\n",kate_get_version_string());
}

int main(int argc,char **argv)
{
  size_t bytes_read;
  int ret=-1;
  const char *input_filename=NULL;
  const char *output_filename=NULL;
  const char *output_filename_type=NULL;
  FILE *fin;
  int verbose=0;
  int arg;
  char signature[64]; /* matches the size of the Kate ID header */
  size_t signature_size;
  int raw;
  char *buffer=NULL;
  ogg_int64_t bytes;
  int headers_written=0;
  int fuzz=0;
  unsigned long fuzz_seed=0;
  ogg_parser_data opd;
  ogg_parser_funcs opf;
  struct lrc_data lrc_data;
  write_funcs wf;

  for (arg=1;arg<argc;++arg) {
    if (argv[arg][0]=='-') {
      switch (argv[arg][1]) {
        case 'V':
          print_version();
          exit(0);
        case 'h':
          print_version();
          printf("usage: %s [options] [filename]\n",argv[0]);
          printf("   -V                  version\n");
          printf("   -v                  verbose\n");
          printf("   -h                  help\n");
          printf("   -o <filename>       set output filename\n");
          printf("   -t <type>           set output format (kate (default), srt, lrc)\n");
          printf("   -B                  write some bitmaps in /tmp (debug)\n");
          printf("   -f <number>         fuzz testing with given seed (debug)\n");
          exit(0);
        case 'o':
          if (!output_filename) {
            output_filename=eat_arg(argc,argv,&arg);
          }
          else {
            fprintf(stderr,"Only one output filename may be converted at a time\n");
            exit(-1);
          }
          break;
        case 'v':
          ++verbose;
          break;
        case 'B':
          write_bitmaps=1;
          break;
        case 'f':
          fuzz=1;
          fuzz_seed=strtoul(eat_arg(argc,argv,&arg),NULL,10);
          break;
        case 't':
          if (!output_filename_type) {
            output_filename_type=eat_arg(argc,argv,&arg);
          }
          else {
            fprintf(stderr,"Only one output type may be given\n");
            exit(-1);
          }
          break;
        default:
          fprintf(stderr,"Invalid option: %s\n",argv[arg]);
          exit(-1);
      }
    }
    else {
      if (!input_filename) {
        input_filename=argv[arg];
      }
      else {
        fprintf(stderr,"Only one input filename may be converted at a time\n");
        exit(-1);
      }
    }
  }

  if (!output_filename_type || !strcmp(output_filename_type,"kate")) {
    wf.write_start_function=&write_kate_start;
    wf.write_headers_function=&write_kate_headers;
    wf.write_end_function=&write_kate_end;
    wf.write_event_function=&write_kate_event;
    wf.write_event_function_data=NULL;
  }
  else if (!strcmp(output_filename_type,"srt")) {
    wf.write_start_function=NULL;
    wf.write_headers_function=NULL;
    wf.write_end_function=NULL;
    wf.write_event_function=&write_srt_event;
    wf.write_event_function_data=NULL;
  }
  else if (!strcmp(output_filename_type,"lrc")) {
    wf.write_start_function=NULL;
    wf.write_headers_function=NULL;
    wf.write_end_function=&write_lrc_end;
    wf.write_event_function=&write_lrc_event;
    wf.write_event_function_data=&lrc_data;
    lrc_data.last_event_end_time=(kate_float)-1.0;
  }
  else {
    fprintf(stderr,"Invalid format type: %s\n",output_filename_type);
    exit(-1);
  }

  fin=open_and_probe_stream(input_filename);
  if (!fin) exit(-1);

  /* first, read the first few bytes to know if we have a raw Kate stream
     or a Kate-in-Ogg stream */
  bytes_read=fread(signature,1,sizeof(signature),fin);
  if (bytes_read!=sizeof(signature)) {
    /* A Kate stream's first packet is 64 bytes, so this cannot be one */
    fprintf(stderr,"Failed to read first %zu bytes of stream\n",sizeof(signature));
    exit(-1);
  }

  if (!memcmp(signature,"\200kate\0\0\0",8)) {
    /* raw Kate stream */
    kate_state k;
    FILE *fout;
    int event_index=0;
    int ret;

    bytes=64;
    buffer=(char*)kate_malloc(bytes);
    if (!buffer) {
      fprintf(stderr,"failed to allocate %lld bytes\n",(long long)bytes);
      exit(-1);
    }
    memcpy(buffer,signature,bytes);

    if (!output_filename || !strcmp(output_filename,"-")) {
      fout=stdout;
    }
    else {
      char *filename=get_filename(output_filename,NULL,NULL);
      fout=fopen(filename,"w");
      if (!fout) {
        fprintf(stderr,"%s: %s\n",filename,strerror(errno));
        exit(-1);
      }
      kate_free(filename);
    }

    raw=1;

    ret=kate_high_decode_init(&k);
    if (ret<0) {
      fprintf(stderr,"failed to init raw kate packet decoding (%d)\n",ret);
      exit(-1);
    }

    while (1) {
      const kate_event *ev=NULL;
      kate_packet kp;
      kate_packet_wrap(&kp,bytes,buffer);
      if (fuzz) fuzz_kate_packet(&fuzz_seed,&kp);
      ret=kate_high_decode_packetin(&k,&kp,&ev);
      if (ret<0) {
        fprintf(stderr,"failed to decode raw kate packet (%d)\n",ret);
        exit(-1);
      }
      if (k.ki->probe<0 && !headers_written) {
        if (wf.write_start_function) (*wf.write_start_function)(fout);
        if (wf.write_headers_function) (*wf.write_headers_function)(fout,k.ki,NULL);
        headers_written=1;
      }
      if (ret>0) {
        if (wf.write_end_function) (*wf.write_end_function)(fout);
        break; /* last packet decoded */
      }
      if (ev) {
        if (wf.write_event_function) {
          (*wf.write_event_function)(fout,wf.write_event_function_data,ev,-1,event_index);
        }
        ++event_index;
      }

      /* all subsequent packets are prefixed with 64 bits (signed) of the packet length in bytes */
      ret=read_raw_size_and_packet(fin,&buffer,&bytes);
      if (ret<0) exit(-1);
    }

    ret=kate_high_decode_clear(&k);
    if (ret<0) {
      fprintf(stderr,"kate_high_decode_clear failed: %d\n",ret);
      /* continue anyway */
    }
    kate_free(buffer);

    if (fout!=stdout) fclose(fout);
  }
  else {
    raw=0;
    signature_size=bytes_read;

    init_kate_stream_set(&opd.kate_streams);
    opd.n_streams=0;
    opd.verbose=verbose;
    opd.fuzz=fuzz;
    opd.fuzz_seed=fuzz_seed;
    opd.output_filename=output_filename;
    opd.wf=&wf;

    opf.on_page=&ogg_parser_on_page;
    ret=parse_ogg_stream(fin,signature,signature_size,opf,(kate_uintptr_t)&opd);

    clear_kate_stream_set(&opd.kate_streams);
  }

  if (fin!=stdin) fclose(fin);

  return 0;
}
